hack/make-rules/test.sh

```diff
@@ -250,22 +235,37 @@ runTests() {
   # must make sure the output from PARALLEL runs is not mixed. To achieve this,
   # we spawn a subshell for each PARALLEL process, redirecting the output to
   # separate files.
-
-  printf "%s\n" "${@}" \
-    | xargs -I{} -n 1 -P "${KUBE_COVERPROCS}" \
+  #
+  # Writing to the combined stdout is serialized with flock.
+
+  printf "%s\n" "${@}" >"${KUBE_TEMP}/packages"
+  gotestsum --format="${gotestsum_format}" \
+            --junitfile="${junit_filename_prefix:+"${junit_filename_prefix}.xml"}" \
+            --raw-command \
+            -- \
+    xargs --arg-file="${KUBE_TEMP}/packages" -I{} -n 1 -P "${KUBE_COVERPROCS}" \
     bash -c "set -o pipefail; _pkg=\"\$0\"; _pkg_out=\${_pkg//\//_}; \
-      go test ${goflags[*]:+${goflags[*]}} \
+      out=\"${junit_filename_prefix:-${KUBE_TEMP}/go-test}-\$_pkg_out.stdout\"; \
+      go test -json ${goflags[*]:+${goflags[*]}} \
         ${KUBE_TIMEOUT} \
         -cover -covermode=\"${KUBE_COVERMODE}\" \
         -coverprofile=\"${cover_report_dir}/\${_pkg}/${cover_profile}\" \
         \"\${_pkg}\" \
         ${testargs[*]:+${testargs[*]}} \
-      | tee ${junit_filename_prefix:+\"${junit_filename_prefix}-\$_pkg_out.stdout\"} \
-      | grep \"${go_test_grep_pattern}\"" \
+      > \"\${out}\" \
+      && rc=$? || rc=$?; \
+      flock 1 cat \"\${out}\"; \
```

Comment:
This is Linux-only as far as I know.

If we want to keep this working on Mac OS X, then some flock/shlock (https://www.unix.com/man-page/osx/1/shlock/) helper script will be needed.

Alternatively, I can drop the second commit from this PR and we simply don't do anything about the coverage case.

Reply:
Or write a small Go command around https://github.com/gofrs/flock in `hack/tools` - we already depend on it there.

Reply:
Yes, running tests should keep working on Mac, that's not a reasonable regression.

Reply:
So which approach do you prefer:
- remove the second commit = change nothing when profiling is enabled
- shell wrapper around flock or shlock
- write Go-based helper tool

Reply:
Never mind. We no longer need this xargs workaround.